### DISCUSSION TITLE
Using Tabulated fluid properties without another fluid

### DISCUSSION THREAD

**Discussion by Jake-HW-Miles** [idaholab/moose#28223] [Q&A Modules: Thermal Hydraulics]
Hi there,
I am trying to get tabulated data to work in one of my input files,  however it doesnt seem to work.

I have tried it in a couple ways, see below for one:
```
  [he]
    type = SimpleFluidProperties
  []

  [he_tab]
    type = TabulatedBicubicFluidProperties
    fluid_property_file = '../csv_store/coolant_salt_tab.csv'
    fp = he
    allow_fp_and_tabulation = true

    interpolated_properties = 'density viscosity k cp cv internal_energy'

    out_of_bounds_behavior = set_to_closest_bound
    error_on_out_of_bounds = false

    construct_pT_from_vh = false
    construct_pT_from_ve = false

    # Tabulation range
    temperature_min = 450
    temperature_max = 845
    pressure_min = 100000.0
    pressure_max = 12000000.0

    # Newton parameters
    tolerance = 1e-8
    T_initial_guess = 647.5
    p_initial_guess = 4000000.0
  []
```

I have tried it without the extra fp object, as well as linearfluid properties, but none seem to work. 

Also, the option for specific volume doesnt seem to exist?

_**Note: The fluid is not helium, it is a ternary chloride salt - so I need it to be treated as a liquid**_

**Comment by GiudGiud** [comment]
Hello

We will need the exact error messages you are hitting. 
It should work, especially with the latest rework. 

Did you make sure to use an updated moose from basically yesterday? Because a lot of the parameters you are using were merged this week. 

Guillaume

**Reply by Jake-HW-Miles** [reply]
I just pulled and made from main, so I should have the most updated.

Im getting a couple different ones as I try a variety of things, the latest of which are:

**Method 1:  Just tabulated fluid properties **

... reading the following properties : 'density viscosity k cp cv internal_energy v'
... I have set the name of "v" to be "specifc_volume" also.
```
 *** ERROR ***
 /root/projects/ORCA/models/systems/2_HTGR/5_scaled_power/p4_tabulated_fp_coolant.i:422.3:
 The following error occurred in the FluidProperties 'he' of type TabulatedBicubicFluidProperties.
 
 virtual void TabulatedFluidProperties::p_from_v_e(libMesh::Real, libMesh::Real, libMesh::Real&, libMesh::Real&, libMesh::Real&) const
 No tabulation or fluid property 'fp' object to compute value
```
**Method 2:  Pairing with linearfluidproperties **

... It does NOT run if I leave linear props empty, as is requesting parameters
... it runs with some reference values taken from the csv but fails to converge. 
... There were also a couple other not implemented errors, but this could be the consequence of something else


For context, the system is a 90MWth primary and secondary (open brayton), where the original primary fluid was helium and we are now trying to change it to a system with a ternary coolant salt. It ran well before changing the fluid medium. If you think we are barking up the wrong tree, please let me know!

**Reply by GiudGiud** [reply]
Ok so option 1 needs a ve tabulation file if you want to use with code that requests p_from_ve. 

You can generate that on your own or you can use additional parameters in the TFP (create_ve_tabulation) to do it

In option 2 those calls that are not known to the pT tabulation (the only one you provided) are forwarded to the fp object, which is why it does not error. 

The documentation for TFP is fairly extensive and describes the 3 ways you can use this.

**Reply by GiudGiud** [reply]
Are you using the thermal hydraulics module for this ternary salt loop?

**Reply by Jake-HW-Miles** [reply]
Thank you GG, I will have a look at generating a ve tabulation file in one way or another.

I would prefer to use option one, so will test this and get back to you.

Also,
I am using thm for the salt loop! I can send parts of my input file if it will be helpful

**Reply by GiudGiud** [reply]
Sounds good. Lemme know if you hit more errors. For this object at this point, it's more a matter of inputting different parameters than developing new things

**Reply by Jake-HW-Miles** [reply]
Making progress! However, I cant see why I am getting this error:

> *** ERROR ***
> /root/projects/ORCA/models/systems/2_HTGR/5_scaled_power/p4_gas_dT_25.i:438.3:
> The following parameter error occurred in the FluidProperties 'he' of type TabulatedBicubicFluidProperties.
> 
> fp: No fluid properties or csv data provided for density.

... I am providing density data, so Im not sure why it is saying this.
I have also had to fix internal energy as it doesnt seem to like it being different for different specific volumes?

**Reply by GiudGiud** [reply]
ok so I ll modify the error to make that clear, but is it density(p,T) or density(v,e) that is crashing? and are you providing both in both pT and ve tabulations?

> I have also had to fix internal energy as it doesnt seem to like it being different for different specific volumes?

That does not seem right. It should form a grid in the ve tabulation.
Something like
v1 e1
v1 e2
v1 e3
v2 e1
v2 e2 
v2 e3
v3 e1
...

**Reply by Jake-HW-Miles** [reply]
> ok so I ll modify the error to make that clear, but is it density(p,T) or density(v,e) that is crashing? and are you providing both in both pT and ve tabulations?

Im only providing ve:
```
  [he]
    type = TabulatedBicubicFluidProperties
    fluid_property_ve_file = "../csv_store/coolant_salt_tab.csv"
    interpolated_properties = 'internal_energy pressure temperature density viscosity k c cv cp entropy'
    create_ve_interpolations = true
    create_pT_interpolations = false
    out_of_bounds_behavior = set_to_closest_bound
  []
```

//

But yes, that v1 v2 v3 ... is how my csv is laid out

**Reply by GiudGiud** [reply]
Weird
I can take a look when I get back this afternoon

**Reply by GiudGiud** [reply]
The more you can share with me the better, I ll need to be able to get to the error

**Reply by Jake-HW-Miles** [reply]
I mean I can attach the csv and input file if you like? 
The input file is a modification of Generic gas cooled micro reactor balance of plant with open-air recuperated Brayton cycle from vtb
And the fluid I am trying to use is from https://www.osti.gov/biblio/1866380
I am currently using a script to generate the file, based on the output from a test I did for a similar fluid (I can also attach?)

**Reply by Jake-HW-Miles** [reply]
We have also had a go at implementing through modifying simplefluidproperties, which seems to run semi well

**Reply by GiudGiud** [reply]
Seems like that would be enough

**Reply by GiudGiud** [reply]
I'll wait for your (reduced) input + csv file here. Thanks for getting in touch for these issues

**Reply by GiudGiud** [reply]
Reminder on this. If you solved it another way we can also close it

**Reply by Jake-HW-Miles** [reply]
Hi, I am sorry I have sort of taken a step back from this to focus on some other work. I will spend sometime recreating the error (as I dont have the original file).

Altering the source code for SimpleFluidProperties was my work around, however would still be good to get this solved!

**Reply by GiudGiud** [reply]
sounds good thanks!
Yes any work on the TFPs is a good investment imo
